### ISSUE TITLE
add scalar() , columns() etc. support to 2.0 Result object.

### ISSUE THREAD

**Issue by rgargente** [sqlalchemy/sqlalchemy#5087]
I often find that I need to transform the results of single column SQLAlchemy queries from a a list of tuples to a list of whatever type the column is. Having an option in SQLA itself would avoid us having take care of this every time.

I wrote code to implement it. The way it would be used is very simple:
```
usernames = s.query(User.name).all(scalar=True)
```
usernames would be a list of strings as opposed of a list of tuples with a single string inside each one.

**Comment by zzzeek** [comment]
hi there -

this feature is already part of a longer term plan in SQLAlchemy 2 and this particular item has not been ticketed yet.     the method will apply to an as-yet-not-created "interim" object that will apply to both Core SELECT statements and similar as well as ORM queries, which by 2.0 will also be via "select()", however in 1.4 the Query object will try to feature some of these new APIs for transitional purposes.

The proposed name at the moment for this method is `.all_scalar()`, but it also needs to be accompanied by a bunch of other result-oriented features, such as being able to control reception of ORM entities alone vs. tuples.  that is, the query.all() method should be returning tuples in **all** cases, however we can't do that in 1.4 because it would be badly backwards incompatible.

basically there's a redesign of the API going on and this is just one little part of it so I'm holding off on query API changes right now until I get to that part of it.

**Comment by rgargente** [comment]
Well I'm glad to see we will get this functionality one way or another!

**Comment by zzzeek** [comment]
yes, it's part of the master plan to fix up all these use cases

**Comment by zzzeek** [comment]
this will be part of the "futrue" API introduced at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#sqlalchemy-1-x-to-2-0-transition

a bunch of use patterns are illustrated at https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#resultproxy-replaced-with-invocationcontext-which-has-more-refined-methods, where "scalar()" is just one of many column-organizing methods:
```

    result.scalar()  # first col of first row  (warns if additional rows remain?)
    result.scalars()  # iterator of first col of each row
    result.scalars().all()  # same, as a list

    result.columns('a', 'b').<anything>  # limit column tuples
    result.columns(table.c.a, table.c.b)  # using Column (or ORM attribute) objects

    result.columns('b', 'a')  # order is maintained

    # if the result is an ORM result, you could do:
    result.columns(User, Address)   # assuming these are available entities


    # index access and slices ?
    result[0].all()  # same as result.scalars().all()
    result[2:5].all()  # same as result.columns('c', 'd', 'e').all()
```

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**Propose Result as immediate replacement for ResultProxy** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1911

**Comment by sqla-tester** [comment]
**Mike Bayer** referenced this issue:

**Integrate new Result into ORM query** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1920

**Comment by zzzeek** [comment]
this is resolved as of  aded39f  and bbf6448.   you can even pass a session.query() to session.execute() and get the Result back.
